File: test_src/fortran/hierarchical_parallelism/reduction/double_precision/target.f90

program target

    

    implicit none
  
    
    DOUBLE PRECISION :: COUNTER =  0   

    
    
     
    
    !$OMP TARGET    MAP(TOFROM: COUNTER) 



    

    


counter = counter +  1.  

 
     

    !$OMP END TARGET
    

    IF  ( ( ABS(COUNTER - 1) ) > 10*EPSILON( COUNTER   ) ) THEN
        write(*,*)  'Expected 1 Got', COUNTER
        call exit(1)
    ENDIF

end program target

File: test_src/cpp/hierarchical_parallelism/reduction/float/target.cpp
#include <iostream>
#include <limits>
#include <cmath>
#include <stdexcept>




bool almost_equal(float x, float y, int ulp) {

     return std::fabs(x-y) <= std::numeric_limits<float>::epsilon() * std::fabs(x+y) * ulp ||  std::fabs(x-y) < std::numeric_limits<float>::min();

}



void test_target(){

 // Input and Outputs
 

float counter{};

// Main program

#pragma omp target   map(tofrom:counter) 

{




counter += float { 1.0f };



}


// Validation
if ( !almost_equal(counter,float { 1 }, 10)  ) {
    std::cerr << "Expected: " << 1 << " Got: " << counter << std::endl;
    throw std::runtime_error( "target give incorect value when offloaded");
}

}
int main()
{
    test_target();
}
